### PR TITLE
TM-4834-Bureau-Exception-List

### PR DESCRIPTION
**Pull request by cagbakwu**
https://metaphase.atlassian.net/browse/TM-4834
-Added new Admin page: Bureau Exception Access
-Added a  list of bureau users 
-Added Edit button 
-Added a page for Bureau add/remove list.
Screenshot 2023-09-06 at 11 16 56 AM
Screenshot 2023-09-06 at 4 28 48 PM

### CODE REVIEW

**Comment by M00staff** [comment]
so, is this still a work in progress or good to be reviewed?

if it's still a WiP - then what is left to be done?

**Comment by M00staff** [comment]
If you want someone to look over something specific - please comment it out in the PR.

**Comment by M00staff** [comment]
Going forward - reach out to Tim before starting work on a ticket, and make sure you understand all of the asks.  

When wrapping up coding, reach out to him again - and do a screen share of what you have locally, so he can go over it with you, to make sure everything is there and looks right... that way we can spend more time reviewing the code, and less time making sure it works.

File: src/Components/AdministratorPage/BureauExceptionList/BureauExceptionListCard.jsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+  return (
+    <div className="position-form">
+      <Row fluid className="bureau-card box-shadow-standard">
+        <Row fluid className="bs-card--row">
+          <Column>
+            Person: {Name}
+          </Column>
+          <Column>
+            Bureau Access: {BureauNames.toString()}
```
Comment (M00staff) [inline src/Components/AdministratorPage/BureauExceptionList/BureauExceptionListCard.jsx:42]: add a space after the comma separation
image

File: src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import Alert from 'Components/Alert';
+import { usePrevious } from 'hooks';
+import { filtersFetchData } from 'actions/filters/filters';
+import { bureauExceptionFetchData, saveBureauExceptionSelections } from 'actions/bureauException';
+import BureauExceptionListCard from './BureauExceptionListCard';
+
+
+const BureauExceptionList = (props) => {
+  const dispatch = useDispatch();
+  const { isAO } = props;
```
Comment (M00staff) [inline src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx:17]: So, I'm not sure if this prop needs to be here ... it looks like you pass it into the `<BureauExceptionListCard />` component, but then don't use it for anything?

image
image

File: src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+  const genericFiltersIsLoading = useSelector(state => state.filtersIsLoading);
+  const userSelections = useSelector(state => state.bureauExceptionsSelections);
+  const BureauExceptionDataLoading = useSelector(state => state.bureauExceptionsFetchDataLoading);
+  const BureauExceptionData = useSelector(state => state.bureauExceptions);
+  const BureauExceptionError = useSelector(state => state.bureauExceptionsFetchDataErrored);
+  const genericFilters = useSelector(state => state.filters);
```
Comment (M00staff) [inline src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx:24]: why are you pulling in generic filters, when there are no filters to filter?

i think you need to be a little more targeted and precise in your copy pasting

Comment (M00staff) [reply src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx:24]: on top of that - turns out, with these new pages we don't even use the generic filters (at least not for any of the ones I've seen)

File: src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+  // Pagination
+  const [page, setPage] = useState(userSelections.page || 1);
+  const prevPage = usePrevious(page);
+  const currentInputs = {
+    page,
+  };
```
Comment (M00staff) [inline src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx:31]: yea - AFAIK, we aren't going to be getting pagination for any of the new pages ... so you can remove all the pagination related code. 

we'll probably have to add a scroll bar and maybe think through a solution for the load times

File: src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx (hunk 3 above)
Comment (M00staff) [reply src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx:24]: we should probably have a filter for names 🤔

you could have the frontend handle it, based on the initial data fetch - i did something similar last night in my Bid Seasons API branch on this repo

File: src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+  const getCurrentInputs = () => ({
+    page,
+  });
+
+  const getQuery = () => ({
+    'bureau-exception-list': true,
+  });
```
Comment (M00staff) [inline src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx:39]: yea - so when you're copying stuff over at light speed, try to ask yourself `"what is this piece of code doing?"`

an example from the Search Post Access Page
image
this bit is normally passed to the API to filter the results, so it calls another search, but with the results filtered ... 

but in your case - there are no filters... 

even if we add the ability to filter names on the frontend, i don't THINK you'll need this at all, and the dispatch function doesn't need an arg.

File: src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    } else if (noResults) {
+      overlay = <Alert type="info" title="No results found" messages={[{ body: 'Please broaden your search criteria and try again.' }]} />;
+    } else {
+      return false;
+    }
+    return overlay;
+  };
+
+  return (
+    genericFiltersIsLoading ?
```
Comment (M00staff) [inline src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx:85]: this is what originally caught my eye - we wait for the generic filters to load before rendering the page ... but, the filters aren't used for anything 😂

File: src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+              </div>
+            </div>
+          </>
+        }
+      </div>
+  );
+};
+
+BureauExceptionList.propTypes = {
+  isAO: PropTypes.bool,
```
Comment (M00staff) [inline src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx:117]: but, why?

File: src/Components/AdministratorPage/BureauExceptionList/EditBureauExceptionList.jsx [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+  const [checkedBureauIds, setCheckedBureauIds] = useState([]);
+  const [bureau, setBureau] = useState('');
+  const [bureauUser, setBureauUser] = useState(user);
+
+  const submit = (e) => {
+    // bureauUser stores the user and bureaus to be added
+    // Doing nothing for now but closing.
+    e.preventDefault();
+    swal.close();
+  };
```
Comment (M00staff) [inline src/Components/AdministratorPage/BureauExceptionList/EditBureauExceptionList.jsx:42]: you could save the next guy some time by passing the action down to this component through its props (or however) and firing that action ... it'll prolly be a post request and the data object will have the ID of the user and the ID of the Bureau to add

File: src/Components/AdministratorPage/BureauExceptionList/EditBureauExceptionList.jsx [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+      setSelectAll(true);
+      setCheckedBureauIds(
+        data.map(bu => bu.id),
+      );
+      setBureauUser({ ...bureauUser, BureauNames: data.map(bu => bu.bureaus[0]) });
+    } else {
+      setSelectAll(false);
+      setCheckedBureauIds([]);
+    }
+  };
```
Comment (M00staff) [inline src/Components/AdministratorPage/BureauExceptionList/EditBureauExceptionList.jsx:64]: was this a requirement?

Comment (M00staff) [reply src/Components/AdministratorPage/BureauExceptionList/EditBureauExceptionList.jsx:64]: i think the list of bureaus will be really, really long ... i removed it from the post access pages

File: src/Components/AdministratorPage/BureauExceptionList/EditBureauExceptionList.jsx [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    { id: 10, bureaus: ['OO'] },
+    { id: 11, bureaus: ['RR'] },
+    { id: 12, bureaus: ['BO'] },
+    { id: 13, bureaus: ['FS'] },
+    { id: 14, bureaus: ['FX'] },
+    { id: 15, bureaus: ['UL'] },
+    { id: 16, bureaus: ['BX'] },
+    { id: 17, bureaus: ['FY'] },
+    { id: 18, bureaus: ['ZW'] },
+  ];
```
Comment (M00staff) [inline src/Components/AdministratorPage/BureauExceptionList/EditBureauExceptionList.jsx:30]: yea - so, I'm looking at the spreadsheet and this data is going to come from an endpoint. 

so, it needs that same Redux boilerplate and will be a simple get request.

File: src/actions/bureauException.js [hunk 11]
```diff
@@ -0,0 +1,375 @@
...
+
+const dummyDataToReturn = (query) => new Promise((resolve) => {
+  const { limit } = query;
+  resolve({
+    results: dummyData.slice(0, limit),
+    count: dummyData.length,
+    next: null,
+    previous: null,
+  });
+});
```
Comment (M00staff) [inline src/actions/bureauException.js:118]: yea - so the response from the new endpoint is gonna be different ... but, pretty much, no pagination and `results` will be called `data`

File: src/actions/bureauException.js [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+      });
+  };
+}
+
+export function bureauExceptionSelectionsSaveSuccess(result) {
+  return {
+    type: 'BUREAU_EXCEPTION_SELECTIONS_SAVE_SUCCESS',
+    result,
+  };
+}
```
Comment (M00staff) [inline src/actions/bureauException.js:190]: yea - so what selections are we saving here?

I'm doing the API stuff for the Bid Seasons page now, and noticed that there were a few Reducers & Actions that didn't do anything, and weren't used anywhere. 

i think there's a lot of the same thing here.

File: src/reducers/bureauException/bureauException.js [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+}
+export function bureauExceptionsPositionSearchFetchDataLoading(state = false, action) {
+  switch (action.type) {
+    case 'BUREAU_EXCEPTIONS_POSITION_SEARCH_FETCH_IS_LOADING':
+      return action.isLoading;
+    default:
+      return state;
+  }
+}
+export function bureauExceptionsPositionSearch(state = [], action) {
```
Comment (M00staff) [inline src/reducers/bureauException/bureauException.js:51]: what are we searching?

**Review by M00staff — CHANGES_REQUESTED** [review]
yea - so I know we are trying to move fast, and you're still ramping up, but please try to proofread your code before having anyone review it. 

there's a lot of un-necessary lines in here.

Try to understand what each line is doing, before adding it - that should also help when building out the other pages.

**Comment by cagbakwu** [comment]
Making corrections 🚀

File: src/Components/AdministratorPage/BureauExceptionList/EditBureauExceptionList.jsx (hunk 9 above)
Comment (cagbakwu) [reply src/Components/AdministratorPage/BureauExceptionList/EditBureauExceptionList.jsx:64]: Yea its needed for the list of bureau codes, its currently in a scrollview.

**Comment by cagbakwu** [comment]
@M00staff this should be good to go too.

File: src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+  const BureauExceptionData = useSelector(state => state.bureauExceptions);
+  const BureauExceptionOptionsData = useSelector(state => state.bureauExceptionsOptions);
+  const BureauExceptionError = useSelector(state => state.bureauExceptionsFetchDataErrored);
+  const fetchAndSet = () => {
+    dispatch(bureauExceptionFetchData());
+    dispatch(bureauExceptionBureauDataFetchData());
+  };
+
+  useEffect(() => {
+    fetchAndSet(false);
```
Comment (M00staff) [inline src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx:24]:
```suggestion
    fetchAndSet();
```

this function doesn't take any arguments ... you also prolly dont need to even abstract to its own function and can just fire the actions here

File: src/Components/AdministratorPage/BureauExceptionList/BureauExceptionListCard.jsx [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+  })),
+  dispatch: PropTypes.func.isRequired,
+};
+
+BureauExceptionListCard.defaultProps = {
+  id: 0,
+  Name: '',
+  BureauNames: '',
+  BureauExceptionOptionsData: [],
+  dispatch: () => {},
```
Comment (M00staff) [inline src/Components/AdministratorPage/BureauExceptionList/BureauExceptionListCard.jsx:84]: you dont really need to pass the dispatch through this component - you can import it at the top of the file

its does, however - need to be passed down to `<EditBureauExceptionList />` though (cause the way sweet alert works)

kind of a nit pick🤷

File: src/Components/AdministratorPage/BureauExceptionList/EditBureauExceptionList.jsx [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+  const data = BureauExceptionOptionsData;
+  const [selectAll, setSelectAll] = useState(false);
+  const [checkedBureauIds, setCheckedBureauIds] = useState([]);
+  const [bureau, setBureau] = useState('');
+  const [bureauUser, setBureauUser] = useState(user);
+
+  const submit = (e) => {
+    // bureauUser stores the user and bureau to be added
+    // Doing nothing for now but closing.
+    dispatch(saveBureauExceptionSelections(bureauUser));
```
Comment (M00staff) [inline src/Components/AdministratorPage/BureauExceptionList/EditBureauExceptionList.jsx:20]: 🙏

**Review by M00staff — APPROVED** [review]
OK - so this is all passable... a few little nit-picks, but this is good to go. 

Hopefully wont be too much of a pain to map the data we get back from the API to the structure you used 🤔

File: src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx (hunk 4 above)
Comment (M00staff) [reply src/Components/AdministratorPage/BureauExceptionList/BureauExceptionList.jsx]: still missing a scroll bar - add like 20 names to your list and you'll see...

could be the list we get back has like 300 names

**Comment by M00staff** [comment]
also, try to be a little cognizant of the labels and what not you have on your PRs ... this one was still marked as WiP

File: src/Components/AdministratorPage/BureauExceptionList/BureauExceptionListCard.jsx (hunk 15 above)
Comment (cagbakwu) [reply src/Components/AdministratorPage/BureauExceptionList/BureauExceptionListCard.jsx:84]: ok ill remove it

**Pull request merged**